Missing column index information in generated parquet file

Hi,

I have currently a problem with the generated parquet file, that I get under (yet) unknown cases the following error:

```
Caused by: shaded.parquet.org.apache.thrift.protocol.TProtocolException: Required field 'null_pages' was not present! Struct: ColumnIndex(null_pages:null, min_values:[69 74 41 50 53 5F 30 31], max_values:[69 74 41 50 53 5F 30 31], boundary_order:null)
	at org.apache.parquet.format.ColumnIndex.validate(ColumnIndex.java:782)
	at org.apache.parquet.format.ColumnIndex$ColumnIndexStandardScheme.read(ColumnIndex.java:918)
	at org.apache.parquet.format.ColumnIndex$ColumnIndexStandardScheme.read(ColumnIndex.java:818)
	at org.apache.parquet.format.ColumnIndex.read(ColumnIndex.java:722)
	at org.apache.parquet.format.Util.read(Util.java:363)
	... 47 more
```

I analyzed it already a bit and found a difference in the schema.

To check the schema, I have used `pqrs` ( https://github.com/manojkarthick/pqrs ).

In the parquet file which is generated via parquetjs, the `metadata` information is empty.

```
version: 1
num of rows: 2
created by: @dsnp/parquetjs
metadata:
message root {
  OPTIONAL BYTE_ARRAY change_id (UTF8);
  OPTIONAL BYTE_ARRAY status (UTF8);
  OPTIONAL BYTE_ARRAY approval_status (UTF8);
```

In the parquet file which was generated via pyspark, where the metadata is filled:

```
version: 1
num of rows: 4096
created by: parquet-mr version 1.12.2 (build f2610ad5b0d33f2882d1d235f0ecbb70da391aea)
metadata:
  org.apache.spark.version: 3.3.2
  org.apache.spark.sql.parquet.row.metadata: {"type":"struct","fields":[{"name":"change_id","type":"string","nullable":true,"metadata":{}},{"name":"status","type":"string","nullable":true,"metadata":{}},{"name":"approval_status","type":"string","nullable":true,"metadata":{}},]}
message spark_schema {
  OPTIONAL BYTE_ARRAY change_id (STRING);
  OPTIONAL BYTE_ARRAY status (STRING);
  OPTIONAL BYTE_ARRAY approval_status (STRING);
```

Here the snippet which I use to generate the parquet file:

```js
import parquet from '@dsnp/parquetjs'

const schema = new parquet.ParquetSchema({
  change_id: parquet.ParquetFieldBuilder.createStringField(true),
  status: parquet.ParquetFieldBuilder.createStringField(true),
  approval_status: parquet.ParquetFieldBuilder.createStringField(true),
})

const records = [
  {
    "change_id": "C-01",
    "status": "closed",
    "approval_status": "approved",
  },
  {
    "change_id": "C-02",
    "status": "closed",
    "approval_status": "approved",
  },
]

export default async function main() {
  const writer = await parquet.ParquetWriter.openFile(
    schema,
    'change.parquet'
  )

  for (const record of records) {
    await writer.appendRow(record)
  }
  await writer.close()
}
```

### Steps to reproduce

Currently I haven't found the code snippet which produces the error.

Running

```
df = spark.read.parquet('<path_to_parquetjs_generated_file>')
df.show()
``` 

directly in a Jupyter Notebook doesn't trigger the error.
It could be somewhere in our calculations/joins which we call in the original script, but I have to analyze this.


### Expected behaviour

The error isn't shown 🙈

### Any other comments?

Tbh. I'm not sure if the missing metadata is the rootcause for this issue.

In the next days I will try to provide some example files with the relevant python code to trigger the error, but I have to finish my work at first to make sure our customer is happy :) 

----
Update 1:

Current workaround is to have a notebook which reads the parquet file and saves it with a new name:

```
df_change = spark.read.parquet(f'{TEST_DATA_PATH}/change.parquet')
df_change.write.mode("overwrite").parquet(f'{TEST_DATA_PATH}/change_spark.parquet')
```

This solves the issue for now - but not really what I want :D

----
Update 2:

While trying to find someone else with the same issue, I have found the following:
https://repost.aws/questions/QUSdc0Pgo9RtSoHOSBwTi8PQ/hive-cannot-open-split-can-not-read-class-org-apache-parquet-format-columnindex

--- a/lib/writer.ts
+++ b/lib/writer.ts
@@ -634,8 +634,11 @@ async function encodeColumnChunk(pages: Page[], opts: {column: ParquetField, bas
 
   /* compile statistics ColumnIndex and OffsetIndex*/
   let columnIndex = new parquet_thrift.ColumnIndex();
+  columnIndex.null_pages = [];
   columnIndex.max_values = [];
   columnIndex.min_values = [];
+  // Default to unordered
+  columnIndex.boundary_order = 0;
   let offsetIndex = new parquet_thrift.OffsetIndex();
   offsetIndex.page_locations = [];
 
@@ -659,6 +662,8 @@ async function encodeColumnChunk(pages: Page[], opts: {column: ParquetField, bas
       statistics.null_count.setValue(statistics.null_count.valueOf() + (page.statistics.null_count?.valueOf() || 0));
       page.distinct_values.forEach((value: unknown) => distinct_values.add(value));
 
+      // If the number of values and the count of nulls are the same, this is a null page
+      columnIndex.null_pages.push( page.num_values === statistics.null_count.valueOf() );
       columnIndex.max_values.push( encodeStatisticsValue(page.statistics.max_value, opts.column) );
       columnIndex.min_values.push( encodeStatisticsValue(page.statistics.min_value, opts.column) );
     }

--- a/test/statistics.js
+++ b/test/statistics.js
@@ -150,34 +150,50 @@ describe('statistics', async function() {
     const name = await reader.envelopeReader.readColumnIndex('name', row);
     assert.deepEqual(name.min_values, ['apples','banana']);
     assert.deepEqual(name.max_values, ['oranges','banana']);
+    assert.deepEqual(name.null_pages, [false, false]);
+    assert.deepEqual(name.boundary_order, 0);
 
     const quantity = await reader.envelopeReader.readColumnIndex('quantity', row);
     assert.deepEqual(quantity.min_values, [10n, undefined]);
     assert.deepEqual(quantity.max_values, [20n, undefined]);
+    assert.deepEqual(quantity.null_pages, [false, false]);
+    assert.deepEqual(quantity.boundary_order, 0);
 
     const price = await reader.envelopeReader.readColumnIndex('price', row);
     assert.deepEqual(price.min_values, [2.6, 3.2]);
     assert.deepEqual(price.max_values, [4.2, 3.2]);
+    assert.deepEqual(price.null_pages, [false, false]);
+    assert.deepEqual(price.boundary_order, 0)
 
     const day = await reader.envelopeReader.readColumnIndex('day', row);
     assert.deepEqual(day.min_values, [ new Date('2008-11-26'), new Date('2017-11-26') ]);
     assert.deepEqual(day.max_values, [ new Date('2018-03-03'), new Date('2017-11-26') ]);
+    assert.deepEqual(day.null_pages, [false, false]);
+    assert.deepEqual(day.boundary_order, 0)
 
     const finger = await reader.envelopeReader.readColumnIndex('finger', row);
     assert.deepEqual(finger.min_values, [ Buffer.from('ABCDE'), Buffer.from('FNORD') ]);
     assert.deepEqual(finger.max_values, [ Buffer.from('XCVBN'), Buffer.from('FNORD')]);
+    assert.deepEqual(finger.null_pages, [false, false]);
+    assert.deepEqual(finger.boundary_order, 0)
 
     const stockQuantity = await reader.envelopeReader.readColumnIndex('stock,quantity', row);
     assert.deepEqual(stockQuantity.min_values, [ 10n, undefined ]);
     assert.deepEqual(stockQuantity.max_values, [ 50n, undefined ]);
+    assert.deepEqual(stockQuantity.null_pages, [false, false]);
+    assert.deepEqual(stockQuantity.boundary_order, 0)
 
     const stockWarehouse = await reader.envelopeReader.readColumnIndex('stock,warehouse', row);
     assert.deepEqual(stockWarehouse.min_values, [ 'A', undefined ]);
     assert.deepEqual(stockWarehouse.max_values, [ 'x', undefined ]);
+    assert.deepEqual(stockWarehouse.null_pages, [false, false]);
+    assert.deepEqual(stockWarehouse.boundary_order, 0)
 
     const colour = await reader.envelopeReader.readColumnIndex('colour', row);
     assert.deepEqual(colour.min_values, [ 'brown', 'yellow' ]);
     assert.deepEqual(colour.max_values, [ 'yellow', 'yellow' ]);
+    assert.deepEqual(colour.null_pages, [false, false]);
+    assert.deepEqual(colour.boundary_order, 0)
 
     const inter = await reader.envelopeReader.readColumnIndex('inter', row).catch(e => e);
     assert.equal(inter.message,'Column Index Missing');

